File: apps/mobile/src/billing/appleProducts.ts
/** Canonical Apple IAP product IDs (App Store Connect — CheckStation Plans). */

export const APPLE_BUNDLE_ID = "app.checkstation.client";

export const APPLE_PRODUCT_IDS = {
  plusMonthly: "app.checkstation.plus.monthly",
  plusYearly: "app.checkstation.plus.yearly",
  businessMonthly: "app.checkstation.business.monthly",
  businessYearly: "app.checkstation.business.yearly",
} as const;

export type AppleProductId = (typeof APPLE_PRODUCT_IDS)[keyof typeof APPLE_PRODUCT_IDS];

export const APPLE_PRODUCT_ID_LIST: AppleProductId[] = [
  APPLE_PRODUCT_IDS.plusMonthly,
  APPLE_PRODUCT_IDS.plusYearly,
  APPLE_PRODUCT_IDS.businessMonthly,
  APPLE_PRODUCT_IDS.businessYearly,
];

export type ApplePlanKey = "plus" | "business";
export type AppleIntervalKey = "monthly" | "yearly";

const PRODUCT_TO_PLAN: Record<AppleProductId, { plan: ApplePlanKey; interval: AppleIntervalKey; titleKey: string }> = {
  [APPLE_PRODUCT_IDS.plusMonthly]: { plan: "plus", interval: "monthly", titleKey: "Plus Monthly" },
  [APPLE_PRODUCT_IDS.plusYearly]: { plan: "plus", interval: "yearly", titleKey: "Plus Yearly" },
  [APPLE_PRODUCT_IDS.businessMonthly]: { plan: "business", interval: "monthly", titleKey: "Business Monthly" },
  [APPLE_PRODUCT_IDS.businessYearly]: { plan: "business", interval: "yearly", titleKey: "Business Yearly" },
};

export function appleProductMeta(productId: string) {
  return PRODUCT_TO_PLAN[productId as AppleProductId] || null;
}

export function appleProductIdFor(plan: ApplePlanKey, interval: AppleIntervalKey): AppleProductId {
  const entry = Object.entries(PRODUCT_TO_PLAN).find(
    ([, meta]) => meta.plan === plan && meta.interval === interval,
  );
  if (!entry) throw new Error(`No Apple product for ${plan}/${interval}`);
  return entry[0] as AppleProductId;
}

File: backend/billing/apple_products.py
"""Canonical Apple IAP product IDs for CheckStation iOS/iPadOS.

Product IDs match App Store Connect (subscription group: CheckStation Plans).
Do not invent alternate IDs here.
"""

from __future__ import annotations

from billing.catalog import PLAN_BUSINESS, PLAN_PLUS
from billing.exceptions import BillingStateError
from billing.models import BillingInterval

APPLE_BUNDLE_ID = "app.checkstation.client"

PRODUCT_PLUS_MONTHLY = "app.checkstation.plus.monthly"
PRODUCT_PLUS_YEARLY = "app.checkstation.plus.yearly"
PRODUCT_BUSINESS_MONTHLY = "app.checkstation.business.monthly"
PRODUCT_BUSINESS_YEARLY = "app.checkstation.business.yearly"

APPLE_PRODUCT_IDS = frozenset(
    {
        PRODUCT_PLUS_MONTHLY,
        PRODUCT_PLUS_YEARLY,
        PRODUCT_BUSINESS_MONTHLY,
        PRODUCT_BUSINESS_YEARLY,
    }
)

# product_id → (plan, interval)
_APPLE_PRODUCT_MAP = {
    PRODUCT_PLUS_MONTHLY: (PLAN_PLUS, BillingInterval.MONTHLY),
    PRODUCT_PLUS_YEARLY: (PLAN_PLUS, BillingInterval.YEARLY),
    PRODUCT_BUSINESS_MONTHLY: (PLAN_BUSINESS, BillingInterval.MONTHLY),
    PRODUCT_BUSINESS_YEARLY: (PLAN_BUSINESS, BillingInterval.YEARLY),
}


def apple_product_ids() -> tuple[str, ...]:
    return (
        PRODUCT_PLUS_MONTHLY,
        PRODUCT_PLUS_YEARLY,
        PRODUCT_BUSINESS_MONTHLY,
        PRODUCT_BUSINESS_YEARLY,
    )


def plan_interval_for_apple_product(product_id: str) -> tuple[str, str]:
    key = str(product_id or "").strip()
    mapped = _APPLE_PRODUCT_MAP.get(key)
    if mapped is None:
        raise BillingStateError(
            "Unknown Apple product.",
            code="apple_unknown_product",
        )
    return mapped


def apple_product_for_plan_interval(plan: str, interval: str) -> str:
    plan_key = str(plan or "").strip().lower()
    interval_key = str(interval or "").strip().lower()
    for product_id, (mapped_plan, mapped_interval) in _APPLE_PRODUCT_MAP.items():
        if mapped_plan == plan_key and mapped_interval == interval_key:
            return product_id
    raise BillingStateError(
        "No Apple product for the requested plan and interval.",
        code="apple_unknown_product",
    )
